### DISCUSSION TITLE
Sound via USB not working?

### DISCUSSION THREAD

**Discussion by oHa510** [BimmerGestalt/AAIdrive#276] [Q&A]
Hi, at first, awesome project! Was looking for something like this a long time, thank you for your efforts!

So I only want to use this app for spotify integration, I compiled the project with my API ID etc. everything working finde so far, I am able to browse my playslists etc. within the AndroidAutoIdrive app.

So after that I connected the phone to my car via USB with the following setup:
- MY2014
- NBT (ID4)
- BMW Connected 6.4
- Samsung Galaxy S5 mini (Android 6.0.1)

Everything is working fine so far ConnectedDrive Apps are showing up and I'm able to browse my playlists and start a song over the idrive spotify menu. It actually looks like it's playing the song (seconds are counting in idrive screen and inside the phone app it shows an equilizer next to the spotify entry), but there is no sound coming from the car... As I understand with an Android version before Oreo (8.0) this should actually work over USB only?

Maybe I am missing something?

**Comment by hufman** [comment]
That all sounds perfect!

The first thing would be to check that the phone is outputting sound over USB. If you turn up the volume, does it come out of the phone's speakers? Maybe the media volume is set low on the USB output device. On my Nexus5 device, I discovered that the AOSP rom wasn't properly outputting USB audio, while the official rom did work. Is there a sound output selection widget in your OS? Perhaps an [after-market app](https://play.google.com/store/apps/details?id=com.nordskog.LesserAudioSwitch) might help debug or force the audio output device. I believe there's a Developer Option about automatic audio routing through USB, perhaps that setting needs toggling?

Next, you would verify that the car is trying to listen to the app connection for audio. The checkbox in the Media menu should move to the Audioplayer to indicate that the car's audio focus is listening to the Audioplayer app connection. You can use the Advanced Settings to force the audio context, in case the [autodetect logic](https://github.com/hufman/AndroidAutoIdrive/blob/37abee7b643a33ec4576607fa16a309e76d734f9/app/src/main/java/me/hufman/androidautoidrive/carapp/music/MusicAppMode.kt#L82) somehow missed something. You could also try out Connected Classic's official Audioplayer, as a known-good test of the audio context functionality.

Hope this helps! Let me know what you find out!

**Reply by oHa510** [reply]
Hi, thanks for your help.

So I did some testing again and got the following results:

- Normal sound output over the phones speakers is working
- Media Volume is set to 100%
- I use official rom
- I tested the after market sound output app -> in normal phone state it says output over "Speakers", after connecting the car it says "USB DAC"
- There indeed is a developer option for automatic USB audio routing, but it's already in the correct state
- Audioplayer in media Menu has the "checkbox" next to it so should be correct
- I tried advanced option "force audio context"
- I tired Connected Classic Audioplayer, it is also playing a song but no sound coming out of the car

So no luck until now :( Is there a way to test the phones USB audio output? Like connect it to a Windows PC and check somehow?

**Reply by hufman** [reply]
Hmmm! That all sounds right, I wish I could help more. I don't know of any easy way to test the phone apart from the car, this Android Open Accessory V2 protocol only really found traction in the car headunit market. I did find [this comment](https://android.stackexchange.com/questions/159342/after-update-car-radio-stopped-playing-audio-through-usb-from-my-samsung-s5-usi) suggesting that his Galaxy S5 was hindered from playing audio because of a conflicting audio switcher app. I didn't find any other suggestions though.

**Reply by mariopil** [reply]
Hello, I've just started using the app and I have the same issue. When playing music while connected over USB the music comes from phone speaker instead of car audio. I have the correct settings in developer options and set in the app to use audio from phone. I'm using latest MyBMW, idrive 6. My phone is Samsung Galaxy A52. Any idea what could cause the problem? I was using android auto in other car over usb and there were no issues with sound there.
Btw are the app logs stored somewhere?

**Reply by hufman** [reply]
The Samsung Galaxy A52 was released with Android 11, which does not include the AOAv2 audio support that the car requires. The advanced option to force AAIdrive to request audio context over USB merely asks the car to try to listen to USB for audio, but your phone will not try to send AOAv2 audio to the car over USB unless it is old enough (Android 8 Oreo).
Android Auto uses a different protocol to send the video and audio streams, and is not relevant to the BMW connection.
The app logs are readily available through `adb logcat`.

**Reply by mariopil** [reply]
Thanks for the reply. So is there any option to force my phone to send the audio over USB?

**Reply by mariopil** [reply]
Ok, I did some research and went through Q&A here and it seems they only option to send audio is over BT. So I will try next to connect AAIdrive over USB for better performance and play audio over BT :)
